### ISSUE TITLE
CORS- Anywhere

### ISSUE THREAD

**Issue by psalm987** [Shrutimahajan/Google-AutoComplete-TextField-Flutter#7]
Getting this error on the console while using the component.

`Access to XMLHttpRequest at 'https://maps.googleapis.com/maps/api/place/autocomplete/json?input=f&key=<key>&components=country:ng' from origin 'https://api.flutlab.io' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.`

The **key**, in the error message, was changed by me.

**Comment by Shrutimahajan** [comment]
ok.. I will check this and update you

**Comment by federicopolesello** [comment]
Hey guys! I have the same issue.
@Shrutimahajan do you have a fix for this?
@psalm987 In the meanwhile did you resolve in some way?
Thank you very much

**Comment by patrisor** [comment]
@psalm987, @Shrutimahajan, and @federicopolesello, the issue you encountered is related to [Cross-Origin Resource Sharing (CORS)](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS). CORS is a security feature implemented by web browsers to prevent unauthorized access to resources from a different origin. In this context, the CORS issue arises when the library attempts to make HTTP requests to the Google Places API from the client web app.

To overcome the CORS restrictions and enable successful API requests, a common workaround is to use a proxy server that is configured to handle CORS requests and forwards them to the Google Places API. The proxy server acts as an intermediary, adding the necessary CORS headers to the requests, allowing them to pass through the browser's security measures.

With help from [this forum](https://stackoverflow.com/questions/61170862/xmlhttprequest-error-with-flutter-web-using-google-places-api-firebase-hosting), here is the solution I came up with:
- [ ] `cd` into the “google_places_flutter-2.0.5/” project directory. 
- You can find the location of this project directory by going into your “.flutter-plugins” file, and making note of the flutter plugins directory. 
- In my case, my full “google_places_flutter-2.0.5/” directory location was “~/.pub-cache/hosted/pub.dev/google_places_flutter-2.0.5/”.
- [ ] Afterwards, `cd` in the “lib/” directory.
- [ ] Create a duplicate of the “google_places_flutter.dart” file with this command, `cp google_places_flutter.dart google_places_flutter.dart.duplicate`, because you are going to need to edit some lines.
- [ ] Add edits to the “google_places_flutter.dart” file:
- Add `import 'package:flutter/foundation.dart' show kIsWeb;` to the top of the file.
- In the `getLocation()` method, replace this line, `
String url = "https://maps.googleapis.com/maps/api/place/autocomplete/json?input=$text&key=${widget.googleAPIKey}";
`, with `
/// Construct the URL for making a Google Places API autocomplete request
String apiURL = "https://maps.googleapis.com/maps/api/place/autocomplete/json?input=$text&key=${widget.googleAPIKey}";
String proxyURL = "https://cors-anywhere.herokuapp.com/";
String url = kIsWeb ? proxyURL + apiURL : apiURL;
/// Add the custom header to the options
final options = kIsWeb ? Options(headers: { "x-requested-with": "XMLHttpRequest" }) : null;
`.
- In the `getLocation()` method, replace this line, `
Response response = await dio.get(url);
`, with `
/// Send an HTTP GET request to the specified URL
Response response = await dio.get(url, options: options);
`.
- In the `getPlaceDetailsFromPlaceId()` method, replace this line, `
var url = "https://maps.googleapis.com/maps/api/place/details/json?placeid=${prediction.placeId}&key=${widget.googleAPIKey}";
Response response = await Dio().get(url);`, with `/// Construct the URL for making a Google Places API
autocomplete request
String apiURL = "https://maps.googleapis.com/maps/api/place/details/json?placeid=${prediction.placeId}&key=${widget.googleAPIKey}";
String proxyURL = "https://cors-anywhere.herokuapp.com/";
String url = kIsWeb ? proxyURL + apiURL : apiURL;
/// Add the custom header to the options
final options = kIsWeb ? Options(headers: { "x-requested-with": "XMLHttpRequest" }) : null;
/// Send an HTTP GET request to the specified URL
Response response = await Dio().get(url, options: options);
`.
- [ ] Save the edited “google_places_flutter.dart” file.
- [ ] Recompile and run your project with `flutter run`.

From this example, the proxy server I ended up using was "https://cors-anywhere.herokuapp.com/". I have seen people use different proxy servers, but this is what’s worked for me. If you decide to change the proxy server, you can replace it by changing the `proxyURL` variable. Here is the full code: [google_places_flutter.dart.txt](https://github.com/Shrutimahajan/Google-AutoComplete-TextField-Flutter/files/12138308/google_places_flutter.dart.txt)

Hope this helps!

**Comment by w113st** [comment]
@psalm987 @patrisor Do you get the error when you add flutlab.io to the domains approved by the API key? Is it because the flutter run web is deploying to a different domain? I think flutlab.io can modify the preview to host of the same domain to avoid this error, as many APIs will fail

**Comment by w113st** [comment]
Looks like https://maps.googleapis.com/maps/api/place/autocomplete is not available for client-side call from web. If you let us change the proxy, it would help.

Another approach is using [google.maps.places.PlaceAutocompleteElement](https://developers.google.com/maps/documentation/javascript/place-autocomplete-new). This can be [enabled with v=beta&libraries=places](https://pub.dev/packages/google_maps_flutter_web). Not sure how to communicate with Flutter for document JS, but an alternative would be to use their API works from client-side web: https://maps.googleapis.com/maps/api/place/js/AutocompletionService.GetPredictionsJson

**Comment by aykutuludag** [comment]
Hello! I'm using this plugin on Android, iOS and Web of same flutter project with same code. Android and iOS working perfectly but Web I'm getting CORS error. Isn't possible add proxyUrl to construct widget on the Dart side? So we can use this plugin on Flutter Web too.

**Comment by aykutuludag** [comment]
> @psalm987, @Shrutimahajan, and @federicopolesello, the issue you encountered is related to [Cross-Origin Resource Sharing (CORS)](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS). CORS is a security feature implemented by web browsers to prevent unauthorized access to resources from a different origin. In this context, the CORS issue arises when the library attempts to make HTTP requests to the Google Places API from the client web app.
> 
> To overcome the CORS restrictions and enable successful API requests, a common workaround is to use a proxy server that is configured to handle CORS requests and forwards them to the Google Places API. The proxy server acts as an intermediary, adding the necessary CORS headers to the requests, allowing them to pass through the browser's security measures.
> 
> With help from [this forum](https://stackoverflow.com/questions/61170862/xmlhttprequest-error-with-flutter-web-using-google-places-api-firebase-hosting), here is the solution I came up with:
> 
> * [ ]  `cd` into the “google_places_flutter-2.0.5/” project directory.
> * You can find the location of this project directory by going into your “.flutter-plugins” file, and making note of the flutter plugins directory.
> * In my case, my full “google_places_flutter-2.0.5/” directory location was “~/.pub-cache/hosted/pub.dev/google_places_flutter-2.0.5/”.
> * [ ]  Afterwards, `cd` in the “lib/” directory.
> * [ ]  Create a duplicate of the “google_places_flutter.dart” file with this command, `cp google_places_flutter.dart google_places_flutter.dart.duplicate`, because you are going to need to edit some lines.
> * [ ]  Add edits to the “google_places_flutter.dart” file:
> * Add `import 'package:flutter/foundation.dart' show kIsWeb;` to the top of the file.
> * In the `getLocation()` method, replace this line, `String url = "https://maps.googleapis.com/maps/api/place/autocomplete/json?input=$text&key=${widget.googleAPIKey}";`, with `/// Construct the URL for making a Google Places API autocomplete request String apiURL = "https://maps.googleapis.com/maps/api/place/autocomplete/json?input=$text&key=${widget.googleAPIKey}"; String proxyURL = "https://cors-anywhere.herokuapp.com/"; String url = kIsWeb ? proxyURL + apiURL : apiURL; /// Add the custom header to the options final options = kIsWeb ? Options(headers: { "x-requested-with": "XMLHttpRequest" }) : null;`.
> * In the `getLocation()` method, replace this line, `Response response = await dio.get(url);`, with `/// Send an HTTP GET request to the specified URL Response response = await dio.get(url, options: options);`.
> * In the `getPlaceDetailsFromPlaceId()` method, replace this line, ` var url = "https://maps.googleapis.com/maps/api/place/details/json?placeid=${prediction.placeId}&key=${widget.googleAPIKey}"; Response response = await Dio().get(url);`, with `/// Construct the URL for making a Google Places API autocomplete request String apiURL = "https://maps.googleapis.com/maps/api/place/details/json?placeid=${prediction.placeId}&key=${widget.googleAPIKey}"; String proxyURL = "https://cors-anywhere.herokuapp.com/"; String url = kIsWeb ? proxyURL + apiURL : apiURL; /// Add the custom header to the options final options = kIsWeb ? Options(headers: { "x-requested-with": "XMLHttpRequest" }) : null; /// Send an HTTP GET request to the specified URL Response response = await Dio().get(url, options: options); `.
> * [ ]  Save the edited “google_places_flutter.dart” file.
> * [ ]  Recompile and run your project with `flutter run`.
> 
> From this example, the proxy server I ended up using was "https://cors-anywhere.herokuapp.com/". I have seen people use different proxy servers, but this is what’s worked for me. If you decide to change the proxy server, you can replace it by changing the `proxyURL` variable. Here is the full code: [google_places_flutter.dart.txt](https://github.com/Shrutimahajan/Google-AutoComplete-TextField-Flutter/files/12138308/google_places_flutter.dart.txt)
> 
> Hope this helps!

Isn't possible to do those thing on Dart side. For example we're giving our API key from Dart side on the Widget code. Isn't possible to do same with giving proxyURL from Dart side?

**Comment by aykutuludag** [comment]
I'm getting this error on both localhost and live.

![image](https://github.com/Shrutimahajan/Google-AutoComplete-TextField-Flutter/assets/18382503/64ad95f3-3ddb-4789-9b8d-b85c1ffa21f4)
![image](https://github.com/Shrutimahajan/Google-AutoComplete-TextField-Flutter/assets/18382503/1f63b35e-8365-4808-871a-094fdcd3c852)

**Comment by monikamatis** [comment]
**Title:** 
Security Vulnerability: `cors-anywhere` Dependency Exposes API Keys in Web Builds

**Issue Type:** 
Bug

**Description:**

The `google_places_flutter` package uses `cors-anywhere` for web requests, exposing the Google Maps Places API key. This is a critical security vulnerability.

**Impact:**

* API Key Theft
* Quota Exhaustion
* Abuse

**Steps to Reproduce:**

1. Use the package in a Flutter web app.
2. Observe network requests in browser dev tools.

**Expected Behavior:**

API key should never be exposed in client-side web code.

**Actual Behavior:**

API key is exposed due to `cors-anywhere`.

**Proposed Solution:**

Implement a server-side proxy.

**Workaround for Web:**

Developers must implement their own server-side proxy.

**Additional Information:**

* Severity: High
* Platform: Web

[https://github.com/Shrutimahajan/Google-AutoComplete-TextField-Flutter/issues/69)]
